From 4af6efe097f0b2755e717b163d9d142d8f133b93 Mon Sep 17 00:00:00 2001
From: Wei Chen <wei.chen@arm.com>
Date: Tue, 15 Aug 2023 20:54:51 +0800
Subject: [PATCH] rss/driver: Add GIC multiple views support for Kronos

Kronos has a single GIC with multiple views for 3 Safety
Island clusters.
1. View-0 is used as a configuration view to configure PEs
   and SPIs for View-1/2/3.
2. All PEs and SPIs belong to Safety Island cluster-0 will
   be assigned to View-1.
3. All PEs and SPIs belong to Safety Island cluster-1 will
   be assigned to View-2.
4. All PEs and SPIs belong to Safety Island cluster-2 will
   be assigned to View-3.

Upstream-Status: Pending
Signed-off-by: Wei Chen <wei.chen@arm.com>
Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Signed-off-by: Robbie Cao <robbie.cao@arm.com>
---
 .../ext/target/arm/rss/common/CMakeLists.txt  |   1 +
 .../common/native_drivers/fainlight_gic_drv.c | 173 +++++++++++++++
 .../common/native_drivers/fainlight_gic_drv.h |  56 +++++
 .../common/native_drivers/fainlight_gic_reg.h |  47 ++++
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |   1 +
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  |  40 ++++
 .../target/arm/rss/kronos/fainlight_gic_lib.c | 209 ++++++++++++++++++
 .../target/arm/rss/kronos/fainlight_gic_lib.h |  17 ++
 .../target/arm/rss/kronos/host_base_address.h |  12 +
 9 files changed, 556 insertions(+)
 create mode 100644 platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.c
 create mode 100644 platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.h
 create mode 100644 platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_reg.h
 create mode 100644 platform/ext/target/arm/rss/kronos/fainlight_gic_lib.c
 create mode 100644 platform/ext/target/arm/rss/kronos/fainlight_gic_lib.h

diff --git a/platform/ext/target/arm/rss/common/CMakeLists.txt b/platform/ext/target/arm/rss/common/CMakeLists.txt
index cb859f554..48900d6a4 100644
--- a/platform/ext/target/arm/rss/common/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/common/CMakeLists.txt
@@ -257,6 +257,7 @@ target_sources(platform_bl2
         $<$<BOOL:${RSS_DEBUG_UART}>:${PLATFORM_DIR}/ext/target/arm/drivers/usart/cmsdk/uart_cmsdk_drv.c>
         native_drivers/lcm_drv.c
         native_drivers/tower_nci_drv.c
+        native_drivers/fainlight_gic_drv.c
         otp_lcm.c
         nv_counters.c
         $<$<BOOL:${RSS_XIP}>:${CMAKE_CURRENT_SOURCE_DIR}/bl2/sic_boot.c>
diff --git a/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.c b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.c
new file mode 100644
index 000000000..779d29f14
--- /dev/null
+++ b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.c
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#include "bootutil/bootutil_log.h"
+#include "fainlight_gic_drv.h"
+#include "fainlight_gic_reg.h"
+
+#include <stdint.h>
+
+/* Definitions to access GICD and GICR registers. */
+#define GICD_READ_REG(dev, ofs) \
+    *(volatile uint32_t *)((dev)->gicd_base + (ofs))
+
+#define GICD_WRITE_REG(dev, ofs, val) \
+    (*(volatile uint32_t *)((dev)->gicd_base + (ofs))) = (uint32_t)(val)
+
+#define GICR_READ_REG(dev, cpu, ofs) \
+    *(volatile uint32_t *) \
+    ((dev)->gicr_base + (dev)->rdist_sz * (cpu) + (ofs))
+
+#define GICR_WRITE_REG(dev, cpu, ofs, val) \
+    (*(volatile uint32_t *) \
+    ((dev)->gicr_base + (dev)->rdist_sz * (cpu) + (ofs))) = (uint32_t)(val)
+
+/* Definitions to calculate SPI register index and bit offset. */
+#define SPI_TO_IVIEWRn_IDX(spi)     (((spi) - 32) >> 4)
+#define SPI_TO_IVIEWRnE_IDX(spi)    (((spi) - 4096) >> 4)
+#define SPI_TO_IVIEWRn_BIT(spi)     (((spi) % 16) << 1)
+#define SPI_IS_VALID(spi)           ((spi) >= 32 && (spi) < 1020)
+#define SPI_IS_VALID_EXT(spi)       ((spi) >= 4096 && (spi) < 5120)
+
+static int assign_pe_to_view(struct gic_mv_dev_t *dev,
+                             uint32_t mpid, uint8_t view)
+{
+    uint32_t val;
+    uint32_t rdist_id;
+
+    if (view >= dev->view_num) {
+        BOOT_LOG_ERR("GIC: Error assigning PE to invaild View:%d", view);
+        return -1;
+    }
+
+    for (rdist_id = 0; rdist_id < dev->rdist_num; rdist_id++) {
+        val = GICR_READ_REG(dev, rdist_id, GICR_TYPER + 4);
+        if (val == mpid)
+            break;
+    }
+
+    if (rdist_id >= dev->rdist_num) {
+        BOOT_LOG_ERR("GIC: Error could not find re-dist for PE:%x", mpid);
+        return -1;
+    }
+
+    /* Update MPID corresponding GICR_VIEWR to view-id */
+    GICR_WRITE_REG(dev, rdist_id, GICR_VIEWR, view);
+
+    return 0;
+}
+
+static int assign_spi_to_view(struct gic_mv_dev_t *dev,
+                              uint16_t spi, uint8_t view)
+{
+    uint32_t val, offset;
+    uint32_t nreg, nbit;
+
+    if (view >= dev->view_num) {
+        BOOT_LOG_ERR("GIC: Error assigning SPI to invaild View:%d", view);
+        return -1;
+    }
+
+    if (SPI_IS_VALID(spi)) {
+        nreg = SPI_TO_IVIEWRn_IDX(spi);
+        offset = GICD_IVIEWnR + 4 * nreg;
+    } else if (SPI_IS_VALID_EXT(spi)) {
+        nreg = SPI_TO_IVIEWRnE_IDX(spi);
+        offset = GICD_IVIEWnRE + 4 * nreg;
+    } else {
+        BOOT_LOG_ERR("GIC: Error assigning invalid SPI:5d to View:%d", spi, view);
+        return -1;
+    }
+    nbit = SPI_TO_IVIEWRn_BIT(spi);
+
+    /* Update SPI corresponding GICD_VIEWR to view-id */
+    val = GICD_READ_REG(dev, offset);
+    val &= (~(GICR_VIEWR_MASK << nbit));
+    val |= (view << nbit);
+    GICD_WRITE_REG(dev, offset, val);
+
+    return 0;
+}
+
+int gic_multiple_view_config_pe(struct gic_mv_dev_t *dev,
+                                struct gic_mv_p2v_map *pe_map,
+                                uint32_t num)
+{
+    int ret;
+    uint32_t idx;
+
+    for (idx = 0; idx < num; idx++) {
+        ret = assign_pe_to_view(dev, pe_map[idx].mpid, pe_map[idx].view);
+        if (ret)
+            return ret;
+    }
+
+    return 0;
+}
+
+int gic_multiple_view_config_spi(struct gic_mv_dev_t *dev,
+                                 struct gic_mv_i2v_map *spi_map,
+                                 uint32_t num)
+{
+    int ret;
+    uint32_t idx, val;
+
+    for (idx = 0; idx < num; idx++) {
+        ret = assign_spi_to_view(dev, spi_map[idx].spi, spi_map[idx].view);
+        if (ret)
+            return ret;
+    }
+
+    /*
+     * The final GICD_CTLR result of each view is the LOGICAL AND of
+     * the GICD_CTLR from view-0 and the GICD_CTLR from each view.
+     * Therefore in the multiview initialization, set the EnableGrp0 and
+     * EnableGrp1 of view-0's GICD_CTLR from firmware so that the software
+     * in each view can control the EnableGrp from their side.
+     * Note that in Safety Island, there is only a single security state.
+     */
+    val = GICD_READ_REG(dev, GICD_CTLR);
+    val |= 0x03;
+    GICD_WRITE_REG(dev, GICD_CTLR, val);
+
+    return 0;
+}
+
+int gic_multiple_view_device_probe(struct gic_mv_dev_t *dev,
+                                   uintptr_t base_addr,
+                                   uint16_t view_num,
+                                   uint16_t rdist_num)
+{
+    uint32_t val;
+
+    dev->view_num  = view_num;
+    dev->rdist_num = rdist_num;
+    dev->gicd_base = base_addr;
+    dev->gicr_base = base_addr + SI_GICR_OFFSET;
+
+    /* Check the GIC version to decide GICR frame size. */
+    val = GICD_READ_REG(dev, GICD_PIDR2) & GICD_PIDR2_ARCH_MASK;
+    val >>= 4;
+    if (val == GICD_PIDR2_ARCH_GICv3)
+        dev->rdist_sz = SI_GICR_FRAME_SZ_v3;
+    else if (val == GICD_PIDR2_ARCH_GICv4)
+        dev->rdist_sz = SI_GICR_FRAME_SZ_v4;
+    else {
+        BOOT_LOG_ERR("GIC: Un-supported GIC version:%d", val);
+        return -1;
+    }
+
+    /* Read GICD_CFGID bit-53 to detect multiple view feature. */
+    val = GICD_READ_REG(dev, GICD_CFGID + 4);
+    if (!(val & GICD_CFGID_HI_MV)) {
+        BOOT_LOG_ERR("GIC: Could not detect Multiple Views Support!");
+        return -1;
+    }
+
+    BOOT_LOG_INF("GIC: Multiple Views Support detected!");
+
+    return 0;
+}
diff --git a/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.h b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.h
new file mode 100644
index 000000000..5ba393e69
--- /dev/null
+++ b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_drv.h
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __FAINLIGHT_GIC_DRV_H__
+#define __FAINLIGHT_GIC_DRV_H__
+
+#include "fainlight_gic_reg.h"
+#include <stdint.h>
+
+/* Structure for GIC multiple view device */
+struct gic_mv_dev_t {
+    uint16_t  view_num;
+    uint16_t  rdist_num;
+    uint32_t  rdist_sz;
+    uintptr_t gicd_base;
+    uintptr_t gicr_base;
+};
+
+/* Structure for GIC Multiple View PE to View mapping */
+struct gic_mv_p2v_map {
+    uint32_t mpid;
+    uint8_t  view;
+};
+
+/* Structure for GIC Multiple View SPI to View mapping */
+struct gic_mv_i2v_map {
+    uint16_t spi;
+    uint8_t  view;
+};
+
+/*
+ * Probe the GIC device and detect the Multiple View feature.
+ */
+int gic_multiple_view_device_probe(struct gic_mv_dev_t *dev,
+                                   uintptr_t base_addr,
+                                   uint16_t view_num,
+                                   uint16_t rdist_num);
+
+/*
+ * Configure the PEs and Views mapping through a P2V map.
+ */
+int gic_multiple_view_config_pe(struct gic_mv_dev_t *dev,
+                                struct gic_mv_p2v_map *pe_map,
+                                uint32_t num);
+/*
+ * Configure the SPIs and Views mapping through an I2V map.
+ */
+int gic_multiple_view_config_spi(struct gic_mv_dev_t *dev,
+                                 struct gic_mv_i2v_map *spi_map,
+                                 uint32_t num);
+
+#endif /* __FAINLIGHT_GIC_DRV_H__ */
diff --git a/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_reg.h b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_reg.h
new file mode 100644
index 000000000..8212b9cd7
--- /dev/null
+++ b/platform/ext/target/arm/rss/common/native_drivers/fainlight_gic_reg.h
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __FAINLIGHT_GIC_REG_H__
+#define __FAINLIGHT_GIC_REG_H__
+
+#include <stdint.h>
+
+/* GIC components offset in each view */
+#define SI_GICD_OFFSET      0x0
+#define SI_GICR_OFFSET      0x40000
+#define SI_GICR_FRAME_SZ_v3 0x20000
+#define SI_GICR_FRAME_SZ_v4 0x40000
+
+/* Register offset of GICD */
+#define GICD_CTLR           0x0
+#define GICD_CFGID          0xF000
+#define GICD_IVIEWnR        0xF600
+#define GICD_IVIEWnRE       0xF800
+#define GICD_PIDR2          0xFFE8
+
+/* GICD_PIDR2 to check the GIC version */
+#define GICD_PIDR2_ARCH_MASK  0xF0
+#define GICD_PIDR2_ARCH_GICv3 0x3
+#define GICD_PIDR2_ARCH_GICv4 0x4
+
+/*
+ * GICD_CFGID MV bit [53] indicates Multiple View support.
+ * Define a bit-21 macro for high 32-bit to check this feature.
+ */
+#define GICD_CFGID_HI_MV     (0x1 << 21)
+
+/* Register offset of GICR */
+#define GICR_TYPER          0x0008
+#define GICR_VIEWR          0x002C
+#define GICR_PIDR2          GICD_PIDR2
+
+#define GICR_VIEWR_MASK     0x3
+
+/* Bit to check the last Redistrubitor on this socket */
+#define GICR_TYPER_LAST     (1U << 4)
+
+#endif /* __FAINLIGHT_GIC_REG_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index 9c768a739..a087bd93d 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -109,6 +109,7 @@ target_sources(platform_bl2
         ${CMAKE_CURRENT_LIST_DIR}/../common/mscp_mailbox.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011/uart_pl011_drv.c
         tfm_interrupts.c
+        fainlight_gic_lib.c
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata/spi_strataflashj3_flash_lib.c
 )

diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index ae9e54a66..568513884 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -27,6 +27,7 @@
 #include "size_defs.h"
 #include "tfm_boot_status.h"
 #include "tfm_plat_defs.h"
+#include "fainlight_gic_lib.h"
 #ifdef CRYPTO_HW_ACCELERATOR
 #include "crypto_hw.h"
 #include "fih.h"
@@ -199,12 +200,51 @@ int boot_store_measurement(uint8_t index,
                              lock_measurement);
 }

+static int32_t gic_multiple_view_init(void)
+{
+    enum atu_error_t atu_err;
+    uint32_t err;
+
+    atu_err = atu_initialize_region(&ATU_DEV_S, RSS_ATU_FAINLIGHT_GIC_ID,
+                                    SI_GIC_VIEW_0_BASE_S_LOG,
+                                    SI_GIC_VIEW_0_BASE_S_PHY,
+                                    SI_GIC_VIEW_SIZE);
+    if (atu_err != ATU_ERR_NONE) {
+        return 1;
+    }
+
+    err = gic_multiple_view_probe(SI_GIC_VIEW_0_BASE_S_LOG);
+    if (err != 0) {
+        BOOT_LOG_INF("BL2: Error probing GIC Multiple Views device");
+        goto free_atu;
+    }
+
+    err = gic_multiple_view_programming();
+    if (err != 0) {
+        BOOT_LOG_INF("BL2: Error programming GIC Multiple Views");
+        goto free_atu;
+    }
+
+free_atu:
+    atu_err = atu_uninitialize_region(&ATU_DEV_S, RSS_ATU_FAINLIGHT_GIC_ID);
+    if (atu_err != ATU_ERR_NONE) {
+        return 1;
+    }
+
+    return 0;
+}
+
 int32_t boot_platform_init(void)
 {
     int32_t result;

     read_chip_id();

+    result = gic_multiple_view_init();
+    if (result != ARM_DRIVER_OK) {
+        return 1;
+    }
+
     result = FLASH_DEV_NAME.Initialize(NULL);
     if (result != ARM_DRIVER_OK) {
         return 1;
diff --git a/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.c b/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.c
new file mode 100644
index 000000000..77e84150d
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.c
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#include "array.h"
+#include "bootutil/bootutil_log.h"
+#include "fainlight_gic_drv.h"
+#include "fainlight_gic_lib.h"
+
+#include <stddef.h>
+
+/* Kronos has view-0/1/2/3 for Safety Island */
+#define GIC_MV_MAX_VIEW_NUM     4
+/*
+ * Kronos GIC View-0 has a contigous region for all
+ * Safety Island GIC re-distributors:
+ * Cluster-0: 1 re-distributors for 1 PE
+ * Cluster-1: 2 re-distributors for 2 PE
+ * Cluster-2: 4 re-distributors for 4 PE
+ */
+#define GIC_MV_MAX_REDIST_NUM  7
+
+static struct gic_mv_dev_t gic_view0;
+
+/* GIC Multiple View, Safety Island Cluster PE to view mapping table */
+static struct gic_mv_p2v_map view_pe_map[] = {
+    /* Cluster 0 PE for View 1 */
+    { 0x0, 1 },
+    /* Cluster 1 PE for View 2 */
+    { 0x10000, 2 },
+    { 0x10100, 2 },
+    /* Cluster 2 PE for View 3 */
+    { 0x20000, 3 },
+    { 0x20100, 3 },
+    { 0x20200, 3 },
+    { 0x20300, 3 }
+};
+
+static struct gic_mv_i2v_map view_spi_map[] = {
+    /* Safety Island CL0 SPI for view 1 */
+    { 34, 1 }, /* System Timer for Cluster 0 */
+    { 37, 1 }, /* 1st interrupt of System Watchdog for Cluster 0 */
+    { 40, 1 }, /* UART for Cluster 0 */
+    { 88, 1 }, /* MHU to PC Cluster 0 NS */
+    { 89, 1 },
+    { 90, 1 },
+    { 91, 1 },
+    { 92, 1 }, /* MHU from PC Cluster 0 NS */
+    { 93, 1 },
+    { 94, 1 },
+    { 95, 1 },
+    { 96, 1 }, /* MHU to PC Cluster 0 S */
+    { 97, 1 },
+    { 98, 1 },
+    { 99, 1 },
+    { 100, 1 }, /* MHU from PC Cluster 0 S */
+    { 101, 1 },
+    { 102, 1 },
+    { 103, 1 },
+    { 104, 1 }, /* MHU to RSS Cluster 0 S */
+    { 105, 1 },
+    { 106, 1 },
+    { 107, 1 },
+    { 108, 1 },
+    { 109, 1 },
+    { 110, 1 },
+    { 111, 1 }, /* MHU from RSS Cluster 0 S */
+    { 112, 1 }, /* MHU CL0 To CL1 Sender */
+    { 114, 1 }, /* MHU CL0 To CL2 Sender */
+    { 117, 1 }, /* MHU CL1 To CL0 Receiver */
+    { 121, 1 }, /* MHU CL2 To CL0 Receiver */
+    { 130, 1 }, /* PC STC 1 */
+    { 131, 1 }, /* PC STC 0 */
+    { 132, 1 }, /* PC PIK  */
+    { 133, 1 }, /* QSPI */
+    { 136, 1 }, /* Ethernet 1 */
+    { 137, 1 }, /* Ethernet 0 */
+    { 138, 1 }, /* DMA 0 */
+    { 141, 1 }, /* ATU Interrupt CL0 */
+    { 143, 1 }, /* NCI main */
+    { 151, 1 }, /* MHU SCP to CL0 receiver */
+    { 152, 1 }, /* MHU CL0 to SCP sender */
+    { 160, 1 }, /* PPU Interrupt for SI CL0 Core 0 */
+    { 164, 1 }, /* PPU Interrupt for SI CL0 */
+    { 165, 1 }, /* PMU Interrupt for SI CL0 */
+
+    /* Safety Island CL1 SPI for view 2 */
+    { 33, 2 }, /* System Timer for Cluster 1 */
+    { 36, 2 }, /* 1st interrupt of System Watchdog for Cluster 1 */
+    { 39, 2 }, /* UART for Cluster 1 */
+    { 72, 2 }, /* MHU to PC Cluster 1 NS */
+    { 73, 2 },
+    { 74, 2 },
+    { 75, 2 },
+    { 76, 2 }, /* MHU from PC Cluster 1 NS */
+    { 77, 2 },
+    { 78, 2 },
+    { 79, 2 },
+    { 80, 2 }, /* MHU to PC Cluster 1 S */
+    { 81, 2 },
+    { 82, 2 },
+    { 83, 2 },
+    { 84, 2 }, /* MHU from PC Cluster 1 S */
+    { 85, 2 },
+    { 86, 2 },
+    { 87, 2 },
+    { 113, 2 }, /* MHU CL0 To CL1 Receiver */
+    { 116, 2 }, /* MHU CL1 To CL0 Sender */
+    { 118, 2 }, /* MHU CL1 To CL2 Sender */
+    { 123, 2 }, /* MHU CL2 To CL1 Receiver */
+    { 128, 2 }, /* FMU */
+    { 129, 2 }, /* SSU */
+    { 134, 2 }, /* FMU Non-Critical */
+    { 139, 2 }, /* DMA 1 */
+    { 145, 2 }, /* ATU Interrupt CL1 */
+    { 153, 2 }, /* MHU SCP to CL1 receiver */
+    { 154, 2 }, /* MHU CL1 to SCP sender */
+    { 168, 2 }, /* PPU Interrupt for SI CL1 Core 0 */
+    { 169, 2 }, /* PPU Interrupt for SI CL1 Core 1 */
+    { 172, 2 }, /* PPU Interrupt for SI CL1 */
+    { 173, 2 }, /* PMU Interrupt for SI CL1 */
+
+    /* Safety Island CL2 SPI for view 3 */
+    { 32, 3 }, /* System Timer for Cluster 2 */
+    { 35, 3 }, /* 1st interrupt of System Watchdog for *nCluster 2 */
+    { 38, 3 }, /* UART for Cluster 2 */
+    { 48, 3 }, /* MHU to PC Cluster 2 NS */
+    { 49, 3 },
+    { 50, 3 },
+    { 51, 3 },
+    { 52, 3 }, /* MHU from PC Cluster 2 NS */
+    { 53, 3 },
+    { 54, 3 },
+    { 55, 3 },
+    { 56, 3 }, /* MHU to PC Cluster 2 S */
+    { 57, 3 },
+    { 58, 3 },
+    { 59, 3 },
+    { 60, 3 }, /* MHU from PC Cluster 2 S */
+    { 61, 3 },
+    { 62, 3 },
+    { 63, 3 },
+    { 64, 3 }, /* MHU to RSS Cluster 2 S */
+    { 65, 3 },
+    { 66, 3 },
+    { 67, 3 },
+    { 68, 3 }, /* MHU from RSS Cluster 2 S */
+    { 69, 3 },
+    { 70, 3 },
+    { 71, 3 },
+    { 115, 3 }, /* MHU CL0 To CL2 Receiver */
+    { 119, 3 }, /* MHU CL1 To CL2 Receiver */
+    { 120, 3 }, /* MHU CL2 To CL0 Sender */
+    { 122, 3 }, /* MHU CL2 To CL1 Sender */
+    { 140, 3 }, /* DMA 2 */
+    { 146, 3 }, /* ATU Interrupt CL2 */
+    { 155, 3 }, /* MHU SCP to CL2 receiver */
+    { 156, 3 }, /* MHU CL2 to SCP sender */
+    { 176, 3 }, /* PPU Interrupt for SI CL2 Core 0 */
+    { 177, 3 }, /* PPU Interrupt for SI CL2 Core 1 */
+    { 178, 3 }, /* PPU Interrupt for SI CL2 Core 2 */
+    { 179, 3 }, /* PPU Interrupt for SI CL2 Core 3 */
+    { 180, 3 }, /* PPU Interrupt for SI CL2 */
+    { 181, 3 }, /* PMU Interrupt for SI CL2 */
+};
+
+int gic_multiple_view_programming(void)
+{
+    int ret;
+
+    BOOT_LOG_INF("GIC: Multiple Views configure PE ...");
+    ret = gic_multiple_view_config_pe(&gic_view0,
+                                      view_pe_map,
+                                      ARRAY_SIZE(view_pe_map));
+    if (ret) {
+        BOOT_LOG_ERR("GIC: configure PE views failed!");
+        return ret;
+    }
+
+    BOOT_LOG_INF("GIC: Multiple Views configure SPI ...");
+    ret = gic_multiple_view_config_spi(&gic_view0,
+                                       view_spi_map,
+                                       ARRAY_SIZE(view_spi_map));
+    if (ret) {
+        BOOT_LOG_ERR("GIC: configure SPI views failed!");
+        return ret;
+    }
+
+    BOOT_LOG_INF("GIC: Multiple Views configuration done!");
+
+    return 0;
+}
+
+int gic_multiple_view_probe(uint32_t view0_base)
+{
+    int ret;
+
+    ret = gic_multiple_view_device_probe(&gic_view0, view0_base,
+                                         GIC_MV_MAX_VIEW_NUM,
+                                         GIC_MV_MAX_REDIST_NUM);
+    if (ret) {
+        BOOT_LOG_ERR("Probe Multiple View GIC device failed!");
+        return ret;
+    }
+
+    return 0;
+}
diff --git a/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.h b/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.h
new file mode 100644
index 000000000..af8857b72
--- /dev/null
+++ b/platform/ext/target/arm/rss/kronos/fainlight_gic_lib.h
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef __FAINLIGHT_GIC_LIB_H__
+#define __FAINLIGHT_GIC_LIB_H__
+
+#include <stdint.h>
+
+int gic_multiple_view_probe(uint32_t view0_base);
+
+int gic_multiple_view_programming(void);
+
+#endif /* __FAINLIGHT_GIC_LIB_H__ */
diff --git a/platform/ext/target/arm/rss/kronos/host_base_address.h b/platform/ext/target/arm/rss/kronos/host_base_address.h
index 50911c191..fb4105f38 100644
--- a/platform/ext/target/arm/rss/kronos/host_base_address.h
+++ b/platform/ext/target/arm/rss/kronos/host_base_address.h
@@ -54,6 +54,8 @@
 #define RSS_ATU_SI_CL2_RSS_MAILBOX_ATU_ID   9 /* ID to use for region mapping to MHU
                                                * outband msg buffer between SI CL2 and RSS
                                                */
+/* ID to use for region mapping to GIC fainlight MMIO pages. */
+#define RSS_ATU_FAINLIGHT_GIC_ID        10

 /* The ATU has a minimum size and all regions are restricted to align with it */
 #define RSS_ATU_PAGE_SIZE           0x2000U /* 8KB */
@@ -313,6 +315,16 @@
 #define HOST_SI_CL2_RSS_MAILBOX_ATU_SIZE   ALIGN_UP(0x1000, RSS_ATU_PAGE_SIZE)
 #define HOST_SI_CL2_RSS_MAILBOX_ATU_ID     RSS_ATU_SI_CL2_RSS_MAILBOX_ATU_ID

+/* Safety Island Multiple View GIC Base addresses */
+#define SI_GIC_VIEW_0_BASE          0x30000000
+#define SI_GIC_VIEW_1_BASE          0x30200000
+#define SI_GIC_VIEW_2_BASE          0x30400000
+#define SI_GIC_VIEW_3_BASE          0x30600000
+#define SI_GIC_VIEW_SIZE            0x200000
+/* Safety Island Multiple View GIC addresses using Non-secure ATU region */
+#define SI_GIC_VIEW_0_BASE_S_LOG   (HOST_ACCESS_BASE_S + 0x8000000)
+/* Safety Island Multiple View GIC addresses */
+#define SI_GIC_VIEW_0_BASE_S_PHY   (HOST_SI_PHYS_BASE + SI_GIC_VIEW_0_BASE)

 /* Safety Island NVM flash layout */
 /*
--
2.25.1

